mcv/patchwork.py

```diff
@@ -0,0 +1,150 @@
+
+PATCHWORK_API_ENDPOINT = os.getenv('PATCHWORK_API_ENDPOINT',
+                                   "https://api.patchworksecurity.com/api/v1")
+DEFAULT_UUID_PATH = '/etc/patchwork/uuid'
+DPKG_QUERY_CMD = ["/usr/bin/dpkg-query", "-W",
+                  "-f", "${Status}\t${Package}\t${Version}\n"]
+
+
+def register(api_key, name, uuid_path=DEFAULT_UUID_PATH):
+    """Register this machine
+
+    UUID metadata is stored in uuid_path. The machine is only
+    registered if the UUID metadata isn't found.
+
+    Returns:
+        dict: JSON metadata for this machine
+    """
+    if os.path.exists(uuid_path):
+        with open(uuid_path, 'r') as f:
+            machine_data = json.load(f)
+
+        return machine_data
+
+    _mkdirs(uuid_path)
+
+    os_data = _get_os_info()
+    os_data.update({'name': name})
+
+    url = "/".join([
+        PATCHWORK_API_ENDPOINT,
+        "machine"])
+
+    machine_data = _post(api_key=api_key, url=url, data=os_data)
+    _save_uuid(uuid_path, machine_data)
+
+    return machine_data
+
+def update(api_key, packages, uuid=None, uuid_path=DEFAULT_UUID_PATH):
+    """Update this machine's package set
+
+    This replaces the package set stored on the server
+
+    Returns:
+        dict: The JSON output from the server
+    """
+    if uuid is None:
+        uuid = _get_uuid(uuid_path)
+
+    data = []
+    
+    for status, package, version in packages:
+        # only upload packages in the 'ii' state otherwise there
+        # may be false positives
+        if status == 'install ok installed':
+            data.append({
+                'name': package,
+                'version': version,
+            })
+
+    url = "/".join([
+        PATCHWORK_API_ENDPOINT,
+        "machine",
+        uuid])
+
+    return _post(api_key=api_key, url=url, data=data)
+
+def gather_packages():
+    """Get the list of packages installed on this machine"""
+    with open(os.devnull, 'w') as devnull:
+        output = subprocess.Popen(DPKG_QUERY_CMD,
+                                  stdout=subprocess.PIPE,
+                                  stderr=devnull).communicate()[0]
+
+    return [line.split('\t') for line in output.splitlines() if line]
+
+def _make_request(api_key, url, method='get', params=None, data=None):
+    """Perform Patchwork API request
+
+    Returns:
+        dict: The JSON output from the server
+
+    Raises:
+        HTTPError: The server returned a 4XX / 5XX error
+    """
+    headers = {'Accept': 'application/json',
+               'Authorization': api_key}
+
+    data = json.dumps(data)
+    r = getattr(requests, method)(
+        url,
+        headers=headers,
+        params=params,
+        data=data,
+        verify=True)
+
+    r.raise_for_status()
+
+    return r.json()
+
+def _mkdirs(uuid_path):
+    """Create missing directories"""
+    dirs = os.path.dirname(uuid_path)
+
+    if not os.path.exists(dirs):
+        os.makedirs(dirs)
+
+def _get_os_info(path='/etc/lsb-release'):
```

Comment:
Ideally this would probably go in like `mcv.distro` or something, I think, so we'd have a general purpose distro identifier function.